src/main/scala/com/microsoft/hyperspace/index/IndexLogEntry.scala

```diff
@@ -29,19 +32,134 @@ case class NoOpFingerprint() {
-  object Directory {
-    // modifiedTime is an Epoch time in milliseconds. (ms since 1970-01-01T00:00:00.000 UTC).
-    case class FileInfo(name: String, size: Long, modifiedTime: Long)
-    object FileInfo {
-      def apply(s: FileStatus): FileInfo =
-        FileInfo(s.getPath.toString, s.getLen, s.getModificationTime)
+  // Create a Content object from a directory path by recursively listing its leaf files. All
+  // files from the directory tree will be part of the Content.
+  def apply(path: Path): Content = Content(Directory(path))
+
+  // Create a Content object from a specified list of leaf files. Any files not listed here will
+  // NOT be part of the returned object
+  def fromLeafFiles(files: Seq[FileStatus]): Content = {
+    /* from org.apache.spark.sql.execution.datasources.PartitionAwareFileIndex. */
+    val leafDirToChildrenFiles: Map[Path, Array[FileStatus]] =
+      files.toArray.groupBy(_.getPath.getParent)
+    val rootPath = getRoot(leafDirToChildrenFiles.head._1)
+
+    val subDirs = leafDirToChildrenFiles
+      .filterNot(_._1.isRoot)
+      .map {
+        case (dir, statuses) => createSubDirectoryTree(dir, statuses)
+      }
+      .toSeq
+
+    val rootFiles: Seq[FileInfo] =
+      leafDirToChildrenFiles.getOrElse(rootPath, Array()).map(FileInfo(_))
+
+    Content(Directory(rootPath.toString, rootFiles, subDirs))
+  }
+
+  // Return file system root path from any path. E.g. "file:/C:/a/b/c" will have root "file:/C:/".
+  // For linux systems, this root will be "file:/". Other hdfs compatible file systems will have
+  // corresponding roots.
+  private def getRoot(path: Path): Path = {
+    if (path.isRoot) path else getRoot(path.getParent)
+  }
+
+  // Create leaf subdirectory. Leaf Subdirectory is the first parent of leaf files.
+  private def createSubDirectoryTree(
+      dirPath: Path,
+      statuses: Array[FileStatus]): Directory = {
+    if (dirPath.getParent.isRoot) {
+      Directory(dirPath.getName, files = statuses.map(FileInfo(_)))
+    } else {
+      createNonLeafSubDirectoryHelper(
+        dirPath.getParent,
+        Directory(dirPath.getName, files = statuses.map(FileInfo(_))))
+    }
+  }
+
+  // Create Non-Leaf subdirectory tree starting from the topmost non-root directory.
+  private def createNonLeafSubDirectoryHelper(dirPath: Path, child: Directory): Directory = {
+    if (dirPath.getParent.isRoot) {
+      Directory(dirPath.getName, files = Seq(), subDirs = Seq(child))
+    } else {
+      createNonLeafSubDirectoryHelper(
+        dirPath.getParent,
+        Directory(dirPath.getName, files = Seq(), subDirs = Seq(child)))
     }
   }
 }
 
+case class Directory(name: String, files: Seq[FileInfo] = Seq(), subDirs: Seq[Directory] = Seq())
+
+object Directory {
+  // Create a Directory object from a directory path by recursively listing its leaf files. All
+  // files from the directory tree will be part of the Content.
+  def apply(path: Path): Directory = {
```

Comment:
Can we rename this to `fromDir` to indicate there are operations done when invoked. Just to clearly differentiate with `Directory("just name")`.

Reply:
thanks @imback82 , done